Field permissions

mattlub opened the pull request:
**UPDATE: ready for review**

relates #134 

what i've done in this PR:
#### 1) add `src/constants/apiPermissions.js` which is empty constant file 

#### 2) move permissions functions to helpers file `helpers/permissions.js`
  - includes new function `getUnauthorizedFields`
  - includes all other functions used by `permissions.js` and `fieldPermissions.js` middleware

#### 3) adds specific field permissioning middleware
  - it checks if user owns resource
  - then finds any unauthorized fields (using `getUnauthorizedFields`, which just uses `req.user.isResourceOwner`)
  - **I updated this from before (last week)**, as before it was shortcutting and just checking if user owned resource by checking `user.isResourceOwner`, because the `permissions` middleware sets it (but only if it checks it). So this would not work for the case where an admin satisfied the minRole condition, and so resource owner is not checked by `permissions`, but then there were fields with `[ SUPER, OWNER ]` permissions in `fieldPermissions`. The resulting middlewares are admittedly not very pretty and probably could be improved greatly by taking a step back. But I will not be doing this unless I return to Naz :)

#### 4) integration tests for `fieldPermissions` using dummy test routes mimicking API route usage

mattlub committed:
add fieldPermissions middleware
#134

mattlub committed:
add empty test file for fieldPermissions

mattlub committed:
fix for lint

mattlub committed:
Merge branch 'apiAuth' into field-permissions

mattlub committed:
Merge branch 'apiAuth' into field-permissions

mattlub committed:
Merge branch 'apiAuth' into field-permissions

mattlub committed:
comment out tests

mattlub committed:
add permissions constants file

mattlub committed:
add permissions to a route

mattlub committed:
add handling for user=undefined in permissions

mattlub committed:
lint fix

mattlub committed:
comment out middleware in api router

mattlub committed:
move middleware into testable function

mattlub committed:
add a test for middleware

mattlub committed:
add tests for getUnauthorizedFields

mattlub committed:
lint

mattlub committed:
lint fix

mattlub committed:
lint fix

mattlub committed:
add permissions constants shell

mattlub committed:
amend fieldPermissions

mattlub committed:
add tests for fieldPermissions initialisation

mattlub committed:
merge master

mattlub committed:
move permissions helpers to own file

mattlub committed:
change fieldPermissions

mattlub committed:
lint fix

mattlub committed:
add tests for fieldPermissions middleware

mattlub committed:
lint fix

mattlub committed:
finish tests for fieldPermissions

mattlub commented:
UPDATE: failing on Travis but not locally, I don't know why

mattlub committed:
add test with POST route

m4v15 commented:
So it's failing on travis on tests that no longer exist, as far as I can tell

m4v15 committed:
case sensitivity test

m4v15 committed:
debug

m4v15 committed:
debugger

mattlub committed:
fix tests by filling out access token properties

mattlub commented:
## @m4v15 YESSSS. I fixed the f*cker.

I'm on linux now, and the tests were failing for me but I could debug. The error was that the expiry date had to be set on the access token, so I've filled out the test tokens with all the attributes. God knows why the tests were passing on our machines before (should we be worried about this?) but all good now.

mattlub commented:
:shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit: :shipit:

m4v15 commented:
Will go through it all tomorrow morning and then hopefully it can get merged - on a side note we now have debug mode on travis if we want it lol

m4v15 committed:
comment out unused vars for linter

will still need them in next PR so not completely losing them

m4v15 committed:
change checkResourceOwner

I've changed it so it resolves to either true or false, instead of rejecting
in the middleware I've changed the usage of it
I think it's neater to reject as unauthorized in the middleware as opposed to
in the helper functions

m4v15 committed:
add comments/documentation

m4v15 committed:
tidy up comments

m4v15 committed:
Merge branch 'master' into field-permissions

m4v15 commented:
@mattlub So it should be relatively clear but what I've done is change the `checkUserOwnsResource` function to just return `true` or `false` (or a promise that resolves to that) instead of this function doing the actual rejection. This allows us to skip the `.catch(() => {})` in fieldPermissions.js which feels messy to me. Actually more importantly (I think) it also means that the function is now just a helper function that does what the name implies - i.e. check the owner, not reject necessarily. Once it is used, we can then decide whether to reject or not.

I've changed the middlewares to make the check and then reject or not depending on the middleware ([reject for permissions](https://github.com/foundersandcoders/open-tourism-platform/blob/8f049aedcfe5e4fc3feb8c3061cd69fc43d3ed7f/src/middleware/permissions.js#L61-L65) and [just set whether owner or not for fieldpermissions](https://github.com/foundersandcoders/open-tourism-platform/blob/8f049aedcfe5e4fc3feb8c3061cd69fc43d3ed7f/src/middleware/fieldPermissions.js#L51))

I think this is much clearer.

Also I've changed the fieldPermissions to only check for the owner if it is an actual resource (which it won't be for POST) - I'm not 100% sure they way I've done this is best (conditional promise chaining is a bit weird) but I think it's ok ([I resolve a Promise and then use an if statement in a `.then`](https://github.com/foundersandcoders/open-tourism-platform/blob/8f049aedcfe5e4fc3feb8c3061cd69fc43d3ed7f/src/middleware/fieldPermissions.js#L45-L53))

N.B. this is why I thought the PR was broke this morning - sorry, it wasn't really I guess, but the conditional chaining and the `.catch(() => {})` meant error handling was getting wierd

m4v15 commented:
@mattlub unless any issues, assign @des-des

mattlub commented:
@m4v15 looks fine

mattlub commented on src/helpers/permissions.js:
```diff
@@ -0,0 +1,115 @@
+  * returns a Promise that resolves to true or false, or rejects if
+  * resourceId is not found in DB (404)
+  *
+  * N.B. when the resource type itself is a User, the method is
```
is this really neessary?

mattlub commented on src/helpers/permissions.js:
```diff
@@ -0,0 +1,115 @@
+
+  return resourceType.findById(resourceId)
+  .then(rejectIfNull(errMessages.GET_ID_NOT_FOUND))
+  .then(doc => {
```
one line?

m4v15 replied:
yeah this was one line originally then I had a bug and forgot to change it back

mattlub commented on src/helpers/permissions.js:
```diff
@@ -0,0 +1,115 @@
+const getUnauthorizedFields = fieldPermissions => fieldsToChange => user => {
+  /*
+  * check an array of the fields a user is trying to change against
+  * an array of those fields and their permission levels
```
is this right? what is "an array of those fields and their permission levels"?

m4v15 replied:
yeah I guess this is phrased badly

mattlub commented on src/helpers/permissions.js:
```diff
@@ -0,0 +1,115 @@
+  return fieldsToChange
+    // filter out fields which are not permissioned
+    .filter(field => permissionedFields.includes(field))
+      // filter down to fields which user is not permitted to change
```
indent

m4v15 replied:
@mattlub you wrote this m8

mattlub replied:
You hijacked my PR!

m4v15 committed:
address lubes' comments

des-des commented:
Woop, ill have a look at this today

des-des commented on src/middleware/fieldPermissions.js:
```diff
@@ -0,0 +1,67 @@
+      if (resourceId) {
+        return checkUserOwnsResource(resourceType)(resourceId)(user)
+        .then(ownerStatus => {
+          req.user.isResourceOwner = ownerStatus
```
Hmm this is just a suggestion, but it might be nice to add this in only once, some middleware before this. (right now being added by 2 different middlewares).

But this kinda stuff can wait until we do a refac somewhere down the line..

I expect this might not be a good idea with the way all the middlewares currently compose together.

m4v15 replied:
Yeah definitely we can refac this later, but I think leave this for now

mattlub replied:
yeh- the decisions was taken because checkUserOwnsResource includes an extra db call, and isn't always necessary. But that would be cleaner for sure.

des-des commented on src/routers/apiRouter.js:
```diff
@@ -2,20 +2,36 @@ const userController = require('../controllers/user')
-const validateJWT = require('../middleware/validateJWT')
-const validateHeaderToken = require('../middleware/validateHeaderToken')
+
+// const validateJWT = require('../middleware/validateJWT')
```
do we want to keep this? If so you should leave a comment explaining

mattlub replied:
I'll just delete it.

des-des commented on src/routers/apiRouter.js:
```diff
@@ -2,20 +2,36 @@ const userController = require('../controllers/user')
   .get(userController.getAll)
-  .post(userController.create)
+  .post(
+    // validateJWT(),
```
same as above

des-des commented on src/routers/apiRouter.js:
```diff
@@ -2,20 +2,36 @@ const userController = require('../controllers/user')
   .put(
-    validateJWT(),
-    validateHeaderToken,
+    // validateJWT(),
```
dito

mattlub replied:
it was kinda half-done already, now it's still half-done but differently.

des-des commented on tests/middleware/fieldPermissions.test.js:
```diff
@@ -0,0 +1,267 @@
+  urlSlice,
+  fieldPermissions(testPermissions),
+  (req, res, next) => res.send('success!'),
+  boomErrorHandler
```
Do you need this?

m4v15 replied:
Yes, we are testing a fake route and some of the tests are to check we get sent back the correct error from this error handler

des-des commented on tests/middleware/fieldPermissions.test.js:
```diff
@@ -0,0 +1,267 @@
+  urlSlice,
+  fieldPermissions(testPermissions),
+  (req, res, next) => res.send('success!'),
+  boomErrorHandler
```
and this?

m4v15 replied:
☝️

des-des reviewed (commented):
Leaving this halfway done, sorry will get to the other half soon!

des-des commented on tests/middleware/fieldPermissions.test.js:
```diff
@@ -0,0 +1,267 @@
+    t.ok(
+      ['f1', 'f2', 'f4', 'f5'].every(field => res.body.message.includes(field)),
+      'message should list fields which are unauthorized')
+    t.notOk(
```
very nice

des-des commented on tests/middleware/fieldPermissions.test.js:
```diff
@@ -0,0 +1,267 @@
+
+tape('fieldPermissions initialization', t => {
+  try {
+    fieldPermissions({
```
test with assertion can be confusing. `t.doesNotThrow`?

mattlub replied:
I didn't know about `t.throws` and `t.doesNotThrow`, that's what I was after.

mattlub replied:
changed

mattlub committed:
clean up routers

mattlub committed:
amend tests to use t.throws and t.doesNotThrow

m4v15 commented:
@mattlub nice work on this, sorry I started to get this done but then other shit got in the way.

des-des commented:
:tada: :sparkles: good work @mattlub @m4v15
